### PR TITLE
WT-6907 Snapshot subpage for Architecture Guide

### PR DESCRIPTION
**Pull request by alisonfel**
Rework the snapshot sub-page entry for the Architecture Guide. This being expanded to discuss the role of snapshots when using transactions, the internals of snapshots and how snapshots are used when checkpointing.

Please let me know if I should add any additional sections on snapshots or expand on any points! I was a bit uncertain with what exact areas should be covered under this topic. Happy to iterate on anything :)

### CODE REVIEW

File: src/docs/arch-snapshot.dox [hunk 1]
```diff
@@ -1,9 +1,169 @@
...
+    however be ignored as its not considered visible at the time of snapshot creation.
+- Minimum transaction ID: The minimum global transaction ID value seen when creating the snapshot. This being the lower
+bound value which all concurrent transactions IDs are greater than or equal to. Transaction IDs strictly less than the minimum value
+are considered visible as these would've been allocated & committed before the snapshot was created. This value is usually taken to be the
+smallest transaction ID value found from the list of concurrent transactions.
+
+An example of building out the aforementioned components of a snapshot can be seen in the below figure. This figure illustrates constructing a snapshot for a transaction with
+ID \c 8, which at the given time, the concurrent running transactions are \c 3, \c 5, \c 7, \c 8 and the global current transaction ID is \c 9.
+
+@plantuml_start{snapshot.png}
```
Comment (alisonfel) [inline src/docs/arch-snapshot.dox:50]: snapshot

File: src/docs/arch-snapshot.dox [hunk 2]
```diff
@@ -1,9 +1,169 @@
...
+
+@section snapshot_transaction Transactions using Snapshots
+
+Within the context of a WiredTiger transaction, a snapshot ensures only versions of committed records
+are visible at the time of the snapshot was created. Transactions more specifically use
+snapshots for the following isolation levels:
+
+- \c snapshot: A snapshot is created at the beginning of a snapshot-level isolated transaction and is sustained throughout
+the entire life cycle of the transaction. In effect, a transaction using \c snapshot isolation will only read versions of
+records before the transaction started. The default isolation level is \c snapshot, and all updates must done at a snapshot
```
Comment (etienneptl) [inline src/docs/arch-snapshot.dox:15]: typo `must *be* done`

**Comment by quchenhao** [comment]
@alisonfel Please have a look at WT-7964 to see if any change there affect the content of this change.

**Comment by alisonfel** [comment]
> @alisonfel Please have a look at [WT-7964](https://jira.mongodb.org/browse/WT-7964) to see if any change there affect the content of this change.

Thanks for the heads up @quchenhao! I based my discussion of `snap_max` on the normal transaction visibility check (that is ID's greater than or equal to `snap_max` are invisible). It seems the linked ticket aligns RTS to be the same as a normal transaction visibility check. I shouldn't need to make any updates :)

File: src/docs/arch-snapshot.dox [hunk 3]
```diff
@@ -1,9 +1,169 @@
...
+1. The transaction ID is greater than or equal to the snapshots recorded maximum transaction ID: If satisfied, the transaction ID is invisible, as the ID
+would've been allocated after the snapshot was created.
+2. The transaction ID is less than the snapshots recorded minimum transaction ID: If satisfied, the transaction ID is visible, as these would've been
+allocated & committed before the snapshot was created.
+3. The transaction ID exists in the snapshots recorded concurrent transaction list: If found, the transaction ID is invisible, as this ID wasn't committed at the
+time the snapshot was created.
+4. Otherwise the ID is visible.
+
+The early ordering of checking whether the transaction ID sits outside the maximum and minimum bounds is an important detail since the concurrent transaction list
+can be empty i.e. snapshot was taken a time when no concurrent transactions where running.
```
Comment (sulabhM) [inline src/docs/arch-snapshot.dox:159]: "snapshot was taken a time when no concurrent transactions where running." --> "snapshot was taken at a time when no concurrent transactions were running."

Notice "a time" --> "at a time" and
"where running" --> "were running"

File: src/docs/arch-snapshot.dox [hunk 4]
```diff
@@ -1,9 +1,169 @@
...
+
+@section snapshot_checkpoint Snapshots \& Checkpoints
+
+Checkpoints require a snapshot to write a transactionally consistent view of the data. A checkpoint maintains it own special global
```
Comment (sulabhM) [inline src/docs/arch-snapshot.dox:163]: "A checkpoint maintains it own special global" --> "A checkpoint maintains its own special global"
i.e.
"it" --> "its"

**Review by sulabhM — APPROVED** [review]
LGTM mod some nit picks.

File: src/docs/arch-snapshot.dox (hunk 2 above)
Comment (alisonfel) [reply src/docs/arch-snapshot.dox]: Good find! Fixed: a880152

File: src/docs/arch-snapshot.dox [hunk 5]
```diff
@@ -1,9 +1,169 @@
...
-When the snapshot isolation level is used in WiredTiger, the transaction
-will see versions of records committed before the transaction started.
+Snapshots are an important mechanism for providing strong isolation guarantees
+in WiredTiger. Snapshots are importantly leveraged by WiredTiger for
+isolating concurrent transactions from reading each others updates.
+
+@section snapshot_transaction Transactions using Snapshots
+
+Within the context of a WiredTiger transaction, a snapshot ensures only versions of committed records
+are visible at the time of the snapshot was created. Transactions more specifically use
```
[the remark this answers is not in the document]
Comment (alisonfel) [reply src/docs/arch-snapshot.dox]: Good point! I used the first suggestion :)
2dd5cd82a8193c80d0fb553edaec7bc1790802c1

File: src/docs/arch-snapshot.dox [hunk 6]
```diff
@@ -1,9 +1,169 @@
...
+Snapshots are implemented by capturing the global state of transactions in the system at the time of snapshot creation. Snapshots
+are managed within per-session transaction contexts, encapsulated by the \c WT_TXN data structure. The main
+data points that encapsulate a given snapshot include:
+
+- Maximum transaction ID: The maximum global transaction ID value seen when creating the snapshot. This being the upper
+bound value which all concurrent transactions IDs are currently smaller than. We resolve the maximum transaction ID value by reading the
+global current transaction ID/counter at the time of snapshot creation. Transactions with IDs greater than or equal to the maximum ID are
+considered invisible to the snapshot.
+- Concurrent transaction IDs: A list of transaction IDs that are concurrently running at the time of the snapshot (i.e. have not
+been committed). As these transactions have yet to be committed at the time of snapshot creation, transactions stored in this list are considered invisible when performing visibility checks.
```
[the remark this answers is not in the document]
Comment (alisonfel) [reply src/docs/arch-snapshot.dox]: I think this is a good point. `are yet` does read a bit better.
Fixed: 5266ad778b83238e904eb9d6d5a79ace1f5f1a85

File: src/docs/arch-snapshot.dox [hunk 7]
```diff
@@ -1,9 +1,169 @@
...
+When building the list of concurrent transactions, WiredTiger walks a global array of running transactions, storing each walked transaction ID in a list managed by the snapshot.
+When walking the global set of running transactions, WiredTiger will ignore a transaction ID that is:
+    -# Its own ID ie. the ID of the transaction creating the given snapshot. A transaction will always read its own updates.
+    -# An ID that is greater than or equal to the snapshots maximum transaction ID. As WiredTiger walks the global list,
```
[the remark this answers is not in the document]
Comment (alisonfel) [reply src/docs/arch-snapshot.dox]: Fixed: 5c0cd8dccea953d8ee55c156f32d4fb97af60174 :)

File: src/docs/arch-snapshot.dox [hunk 8]
```diff
@@ -1,9 +1,169 @@
...
+    it may find a transaction that has concurrently allocated a new transaction ID, exceeding the snapshots maximum ID bounds. The will
+    however be ignored as its not considered visible at the time of snapshot creation.
```
[the remark this answers is not in the document]
Comment (alisonfel) [reply src/docs/arch-snapshot.dox]: Fixed: 5c0cd8d :)

File: src/docs/arch-snapshot.dox [hunk 9]
```diff
@@ -1,9 +1,169 @@
...
+- Minimum transaction ID: The minimum global transaction ID value seen when creating the snapshot. This being the lower
+bound value which all concurrent transactions IDs are greater than or equal to. Transaction IDs strictly less than the minimum value
+are considered visible as these would've been allocated & committed before the snapshot was created. This value is usually taken to be the
```
[the remark this answers is not in the document]
Comment (alisonfel) [reply src/docs/arch-snapshot.dox]: Good point, `and` is better.
Fixed:  cdc744b5b6417195a3dc125a21805972e2b0c3c8

File: src/docs/arch-snapshot.dox [hunk 10]
```diff
@@ -1,9 +1,169 @@
...
+smallest transaction ID value found from the list of concurrent transactions.
+
+An example of building out the aforementioned components of a snapshot can be seen in the below figure. This figure illustrates constructing a snapshot for a transaction with
+ID \c 8, which at the given time, the concurrent running transactions are \c 3, \c 5, \c 7, \c 8 and the global current transaction ID is \c 9.
```
[the remark this answers is not in the document]
Comment (alisonfel) [reply src/docs/arch-snapshot.dox]: Yep, that makes sense. I've divided the sentence into two: 2b874123b37aeab1049a451d21850ee6c8679589

File: src/docs/arch-snapshot.dox (hunk 3 above)
Comment (alisonfel) [reply src/docs/arch-snapshot.dox]: Good find. Fixed: 89b30d80bce222ef88c7f6d2caab9c16d4215201

File: src/docs/arch-snapshot.dox (hunk 4 above)
Comment (alisonfel) [reply src/docs/arch-snapshot.dox]: Fixed! 551a4e12a694b00d764c94568c17c2cb8bce6f22

**Review by etienneptl — APPROVED** [review]
LGTM

**Review by bvpvamsikrishna — APPROVED** [review]
lgtm

**Pull request merged**